File: VSCode_Python/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False
        mapping = {
            '(': ')',
            '[': ']',
            '{': '}'
        }
        stack = []
        for bracket in s:
            if bracket in mapping:
                stack.append(bracket)
            else:
                if not stack or bracket!=mapping[stack[-1]]:
                    return False
                stack.pop()
        return not stack
# @lc code=end

File: VSCode_CPP/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        unordered_map <char, char> mp = {
            {']', '['},
            {'}', '{'},
            {')', '('},
        };
        stack<char> st;
        for (auto x:s) {
            if (mp.count(x) > 0) {
                if (st.empty() || st.top() != mp[x]) return false;
                st.pop();
            } else {
                st.push(x);
            }
        }
        return st.empty();
    }
};
// @lc code=end

